File: services/scbe-shim/src/axioms.ts
/**
 * MVP axiom checks — surface heuristics, NOT the full 14-layer pipeline.
 *
 * Each check returns a violation score in [0, 1] where 0 = clean and
 * 1 = strongly violated. The harmonic score combines all five into the
 * Layer-12 decision: ALLOW / QUARANTINE / ESCALATE / DENY.
 *
 * The full 14-layer pipeline (Poincare embedding, Mobius phase, spectral
 * coherence, etc.) is at src/harmonic/pipeline14.ts in the main repo.
 * This file is the production proxy form — cheap-to-run, suitable for
 * Cloudflare Worker 50ms CPU budget.
 */

export type AxiomName =
  | "unitarity"  // A1 — norm / bounded output
  | "locality"   // A2 — stays within request scope
  | "causality"  // A3 — no time-ordering violations / future leak
  | "symmetry"   // A4 — no asymmetric bias / refusal injection
  | "composition"; // A5 — response shape matches request shape

export interface AxiomViolation {
  axiom: AxiomName;
  score: number;        // [0, 1]
  reason: string;       // short code for telemetry
  detail?: string;      // optional human-readable explanation
}

export interface AxiomReport {
  violations: AxiomViolation[];
  worstScore: number;
  worstAxiom: AxiomName | null;
  reasons: string[];
}

const MAX_REASONABLE_LENGTH = 16_000;
const TOO_SHORT_MIN = 1;

// A1 — Unitarity: the response should be present, bounded, and not empty.
function checkUnitarity(output: string): AxiomViolation | null {
  const len = output.length;
  if (len < TOO_SHORT_MIN) {
    return { axiom: "unitarity", score: 1.0, reason: "axiom:unitarity.empty", detail: "model returned empty content" };
  }
  if (len > MAX_REASONABLE_LENGTH) {
    return { axiom: "unitarity", score: 0.8, reason: "axiom:unitarity.overflow", detail: `output ${len} chars exceeds ${MAX_REASONABLE_LENGTH}` };
  }
  return null;
}

// A2 — Locality: output should not contain telltale exfil patterns
// (system prompt leakage, internal tool name leakage, instructions to
// reach external systems).
function checkLocality(output: string): AxiomViolation | null {
  const exfilPatterns: Array<[RegExp, string]> = [
    [/\b(system prompt|hidden instructions?|developer instructions?)\b/i, "axiom:locality.system_prompt_leak"],
    [/\b(rollback|restart)_conversation\b/i, "axiom:locality.tool_leak"],
    [/\bignore (previous|prior|all|above) instructions?\b/i, "axiom:locality.jailbreak_phrase"],
    [/\bbase64\s*[:=]\s*[A-Za-z0-9+/]{40,}/, "axiom:locality.suspicious_base64"],
  ];
  for (const [re, reason] of exfilPatterns) {
    if (re.test(output)) {
      return { axiom: "locality", score: 0.9, reason };
    }
  }
  return null;
}

// A3 — Causality: no time-ordering violations. MVP check: model shouldn't
// claim to know the user's future actions or pretend the conversation
// happened in a different order.
function checkCausality(output: string): AxiomViolation | null {
  const patterns: Array<[RegExp, string]> = [
    [/\bas you (will|are about to) (say|do|ask)\b/i, "axiom:causality.future_assertion"],
    [/\bI already (told|gave) you\b/i, "axiom:causality.past_assertion_loop"],
  ];
  for (const [re, reason] of patterns) {
    if (re.test(output)) {
      return { axiom: "causality", score: 0.5, reason };
    }
  }
  return null;
}

// A4 — Symmetry: detect canned refusal injection (model was steered into
// refusing without a real grounded reason) AND detect one-sided harmful
// content endorsement. MVP heuristic: balance "I cannot" / "I refuse"
// without any specific reason.
function checkSymmetry(output: string): AxiomViolation | null {
  const lower = output.toLowerCase();
  // Bare refusal with no reason given (we call this asymmetric)
  if (
    /\bi (cannot|can't|won't|refuse to)\b/.test(lower) &&
    !/\bbecause\b|\bdue to\b|\bin order to\b|\bsince\b/.test(lower)
  ) {
    return { axiom: "symmetry", score: 0.4, reason: "axiom:symmetry.bare_refusal" };
  }
  // Endorsement of clearly harmful operative content
  if (/\bhere(?:'s| is) how to (?:make|build|synthesize) (?:a bomb|explosives|methamphetamine|ricin)\b/i.test(output)) {
    return { axiom: "symmetry", score: 1.0, reason: "axiom:symmetry.harmful_endorsement" };
  }
  return null;
}

// A5 — Composition: response shape should roughly match request shape.
// MVP: if the request looked like a chat message but the response is
// only structural noise (code fences with no prose), flag soft.
function checkComposition(output: string, requestUserMessage: string): AxiomViolation | null {
  const trimmed = output.trim();
  if (trimmed.length === 0) {
    return { axiom: "composition", score: 1.0, reason: "axiom:composition.empty_after_trim" };
  }
  // If the user asked a yes/no question but the response is a 4000-char essay,
  // flag soft (composition mismatch).
  const userIsShortQuestion =
    requestUserMessage.trim().length < 80 && /\?$/.test(requestUserMessage.trim());
  if (userIsShortQuestion && trimmed.length > 4000) {
    return { axiom: "composition", score: 0.3, reason: "axiom:composition.length_mismatch" };
  }
  return null;
}

export function evaluateAxioms(output: string, requestUserMessage: string): AxiomReport {
  const checks: Array<AxiomViolation | null> = [
    checkUnitarity(output),
    checkLocality(output),
    checkCausality(output),
    checkSymmetry(output),
    checkComposition(output, requestUserMessage),
  ];
  const violations = checks.filter((v): v is AxiomViolation => v !== null);
  const reasons = violations.map(v => v.reason);
  let worstScore = 0;
  let worstAxiom: AxiomName | null = null;
  for (const v of violations) {
    if (v.score > worstScore) {
      worstScore = v.score;
      worstAxiom = v.axiom;
    }
  }
  return { violations, worstScore, worstAxiom, reasons };
}

File: services/scbe-shim-space/shim/axioms.py
"""Python mirror of services/scbe-shim/src/axioms.ts."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Optional

MAX_REASONABLE_LENGTH = 16_000
TOO_SHORT_MIN = 1


@dataclass
class AxiomViolation:
    axiom: str
    score: float
    reason: str
    detail: Optional[str] = None


@dataclass
class AxiomReport:
    violations: List[AxiomViolation] = field(default_factory=list)
    worst_score: float = 0.0
    worst_axiom: Optional[str] = None
    reasons: List[str] = field(default_factory=list)


def _check_unitarity(output: str) -> Optional[AxiomViolation]:
    n = len(output)
    if n < TOO_SHORT_MIN:
        return AxiomViolation("unitarity", 1.0, "axiom:unitarity.empty", "model returned empty content")
    if n > MAX_REASONABLE_LENGTH:
        return AxiomViolation(
            "unitarity", 0.8, "axiom:unitarity.overflow", f"output {n} chars exceeds {MAX_REASONABLE_LENGTH}"
        )
    return None


def _check_locality(output: str) -> Optional[AxiomViolation]:
    patterns = [
        (
            re.compile(r"\b(system prompt|hidden instructions?|developer instructions?)\b", re.I),
            "axiom:locality.system_prompt_leak",
        ),
        (re.compile(r"\b(rollback|restart)_conversation\b", re.I), "axiom:locality.tool_leak"),
        (re.compile(r"\bignore (previous|prior|all|above) instructions?\b", re.I), "axiom:locality.jailbreak_phrase"),
        (re.compile(r"\bbase64\s*[:=]\s*[A-Za-z0-9+/]{40,}"), "axiom:locality.suspicious_base64"),
    ]
    for rx, reason in patterns:
        if rx.search(output):
            return AxiomViolation("locality", 0.9, reason)
    return None


def _check_causality(output: str) -> Optional[AxiomViolation]:
    patterns = [
        (re.compile(r"\bas you (will|are about to) (say|do|ask)\b", re.I), "axiom:causality.future_assertion"),
        (re.compile(r"\bI already (told|gave) you\b", re.I), "axiom:causality.past_assertion_loop"),
    ]
    for rx, reason in patterns:
        if rx.search(output):
            return AxiomViolation("causality", 0.5, reason)
    return None


def _check_symmetry(output: str) -> Optional[AxiomViolation]:
    lower = output.lower()
    if re.search(r"\bi (cannot|can't|won't|refuse to)\b", lower) and not re.search(
        r"\bbecause\b|\bdue to\b|\bin order to\b|\bsince\b", lower
    ):
        return AxiomViolation("symmetry", 0.4, "axiom:symmetry.bare_refusal")
    if re.search(
        r"\bhere(?:'s| is) how to (?:make|build|synthesize) (?:a bomb|explosives|methamphetamine|ricin)\b",
        output,
        re.I,
    ):
        return AxiomViolation("symmetry", 1.0, "axiom:symmetry.harmful_endorsement")
    return None


def _check_composition(output: str, user_message: str) -> Optional[AxiomViolation]:
    trimmed = output.strip()
    if len(trimmed) == 0:
        return AxiomViolation("composition", 1.0, "axiom:composition.empty_after_trim")
    user_trim = user_message.strip()
    user_is_short_question = len(user_trim) < 80 and user_trim.endswith("?")
    if user_is_short_question and len(trimmed) > 4000:
        return AxiomViolation("composition", 0.3, "axiom:composition.length_mismatch")
    return None


def evaluate_axioms(output: str, user_message: str) -> AxiomReport:
    checks = [
        _check_unitarity(output),
        _check_locality(output),
        _check_causality(output),
        _check_symmetry(output),
        _check_composition(output, user_message),
    ]
    violations = [v for v in checks if v is not None]
    worst_score = 0.0
    worst_axiom: Optional[str] = None
    for v in violations:
        if v.score > worst_score:
            worst_score = v.score
            worst_axiom = v.axiom
    return AxiomReport(
        violations=violations,
        worst_score=worst_score,
        worst_axiom=worst_axiom,
        reasons=[v.reason for v in violations],
    )
